CASSANDRA-18508 Enhancing sensitive JMX SSL Config

maulin-vasavada opened the pull request:
The [Cassandra Jira 18508](https://issues.apache.org/jira/browse/CASSANDRA-18508).

This PR enhances ability to specify JMX SSL configuration via `cassandra.yaml` in the same way it can be done for `Native transport/Internode messaging` using `encryption_options`. This removes the need for specifying sensitive configuration for JMX SSL like key/trust store passwords in `cassandra-env.sh`. This PR also adds required unit and integration tests for JMX SSL configuration covering existing code prior to this change.

Commit:
CASSANDRA-18508 Enhancing sensitive JMX SSL Config

CASSANDRA-18508 Enhancing sensitive JMX SSL Config

smiklosovic commented on conf/cassandra-env.sh:
```diff
@@ -236,6 +236,8 @@ else
   JVM_OPTS="$JVM_OPTS -Dcom.sun.management.jmxremote.authenticate=true"
 
   # jmx ssl options
+  # Consider using the jmx_encryption_options section of cassandra.yaml instead to prevent sensitive information being
```
could we make it like "exposed" is not on a new line alone? Either make it one long line or break it in the middle and create two if you feel like it.

smiklosovic replied:
What if I have both cassandra.yaml as well as cassandra-env.sh configured? What is the precedence? How does this work? It should be possible to configure it at one place only, either here OR in cassandra.yaml and we should error out on startup when this kind of configuration is found at both places. Having it possibly at both places just brings confusion.

maulin-vasavada replied:
If both `cassandra.yaml` and `cassandra-env.sh` are configured the precedence is cassandra-env.sh since that is the prior functionality before this change. You can see this condition check [here](https://github.com/apache/cassandra/pull/3638/files#diff-039489ddb488470626dad5fac18d1541da0bd23f3137178cf3f63b29e020d5daR55) in the code. However, this means we should clarify this in the documentation update as well as in the cassandra-env.sh?

smiklosovic commented on examples/ssl-factory/test/unit/org/apache/cassandra/security/KubernetesSecretsPEMSslContextFactoryTest.java:
```diff
@@ -114,7 +114,7 @@ public void setup()
          * In order to test with real 'env' variables comment out this line and set appropriate env variable. This is
          * done to avoid having a dependency on env in the unit test.
          */
-        commonConfig.put("require_client_auth", "false";
```
ouch

maulin-vasavada replied:
Yes, I was quiet surprised when I could not compile the ssl-example from trunk so this was my first commit before I could start.

maulin-vasavada replied:
Interesting. In my local code I do see that line intact :(

![image](https://github.com/user-attachments/assets/0816ef4c-178d-4869-82a5-d6511978398f)

maulin-vasavada replied:
Oh I recollect now. The trunk was broken for this line when I started these changes. So I added the missing `)` to make it compile. So my local diff shows below and the next line in the PR also shows the same. Hence I don't think there is any issues. 
![image](https://github.com/user-attachments/assets/a00d654b-b143-4092-bf89-db7c2a43f723)

smiklosovic commented on src/java/org/apache/cassandra/config/DatabaseDescriptor.java:
```diff
@@ -3651,12 +3661,23 @@ public static EncryptionOptions getNativeProtocolEncryptionOptions()
     }
 
+    @VisibleForTesting
+    public static void updateJmxEncryptionOptions(Function<EncryptionOptions, EncryptionOptions> update)
```
@maulin-vasavada my IDE tells me that this method is not used anywhere. Where is it used? If not used then it should not be there.

smiklosovic commented on src/java/org/apache/cassandra/config/EncryptionOptions.java:
```diff
@@ -173,7 +173,7 @@ public enum ConfigKey
         }
 
-        String getKeyName()
+        public String getKeyName()
```
@maulin-vasavada Is this really needed? What about leveraging `EncryptionOptions.toString()` where `toString` would return `keyName`? `getKeyName` would not be necessary at all then.

While you are at it, could you fix spaces in  body of the constructor around `=` and remove braces for `for` in `asSet` ? (or at least move the opening one to a new line).

maulin-vasavada replied:
This is mainly used to avoid repeating string key names for the following two classes. We don't really have EncryptionOption object to call `toString()` on in this case. Let me know what do you think.
![image](https://github.com/user-attachments/assets/22071908-a374-4dee-be15-dda6466c9b7a)

maulin-vasavada replied:
While you are deciding on the `getKeyName()` change based on latest comment, I can look at the constructor change for spaces.

maulin-vasavada replied:
I fixed the constructor spaces around `=` and the `for` loop braces.

smiklosovic replied:
@maulin-vasavada so what if we replaced all calls of `getKeyName()` with `toString()` ?

maulin-vasavada replied:
Will do.

smiklosovic commented on src/java/org/apache/cassandra/utils/jmx/DefaultJmxSocketFactory.java:
```diff
@@ -0,0 +1,84 @@
+        env.put(RMIConnectorServer.RMI_SERVER_SOCKET_FACTORY_ATTRIBUTE, new RMIServerSocketFactoryImpl(serverAddress));
+    }
+
+    @Override
```
While I was reading this, I was little bit confused what each `configureSslServerSocketFactory` does. Three methods here are named same, it is just they return something else and what they return is derived from what parameters it takes. I would expect that the methods would be named more descriptively so I know what it is going to return without looking into how it is implemented.

maulin-vasavada replied:
Currently its method overloading with different input params to effectively do the same thing - `configure the relevant ssl server socket factory`. We have two use-cases primarily-

1. setting the socket factory based on system configure key/trust store
2. setting the socket factory based on JMX `encryption_options`. 

The third `configure` method is for Client socket factory.

We have JavaDocs about those in [the parent abstract class](https://github.com/apache/cassandra/pull/3638/files#diff-039489ddb488470626dad5fac18d1541da0bd23f3137178cf3f63b29e020d5da). 

Let me know if it still makes more sense to rename methods to reflect what it does.

smiklosovic replied:
@maulin-vasavada why dont you reflect what it returns in the method names?

smiklosovic commented on src/java/org/apache/cassandra/utils/jmx/AbstractJmxSocketFactory.java:
```diff
@@ -0,0 +1,153 @@
+     * </pre>
+     * @param jmxEncryptionOptions
+     */
+    void setJmxSystemProperties(EncryptionOptions jmxEncryptionOptions)
```
can be `private`.

smiklosovic replied:
@maulin-vasavada still valid

maulin-vasavada replied:
done

smiklosovic commented on src/java/org/apache/cassandra/utils/jmx/AbstractJmxSocketFactory.java:
```diff
@@ -0,0 +1,153 @@
+    void setJmxSystemProperties(EncryptionOptions jmxEncryptionOptions)
+    {
+        COM_SUN_MANAGEMENT_JMXREMOTE_SSL.setBoolean(true);
+        if (jmxEncryptionOptions.getAcceptedProtocols() != null)
```
remove braces for `if`s, they are not necessary.

smiklosovic replied:
@maulin-vasavada

maulin-vasavada replied:
done

smiklosovic commented on src/java/org/apache/cassandra/utils/jmx/IJmxSocketFactory.java:
```diff
@@ -0,0 +1,60 @@
+     * and {@code com.sun.jndi.rmi.factory.socket} properties for the client and server socket factories.
+     * @throws SSLException if it fails to configure the socket factories with the given input
+     */
+    Map<String, Object> configure(InetAddress serverAddress, boolean localOnly, EncryptionOptions jmxEncryptionOptions)
```
I would move this all comment to place where it is implemented. Here you are making assumption what the implementation does. Does not make sense to me to have javadoc here.

maulin-vasavada replied:
When I look at it now, your point makes sense. I'll move the javadocs to the Default implementation and add javadocs for this interface appropriately.

maulin-vasavada replied:
done, please review. I removed the implementation specific docs from the interface method and added the detailed documentation to the abstract class it implements the method with specific details.

smiklosovic commented on test/distributed/org/apache/cassandra/distributed/impl/IsolatedJmx.java:
```diff
@@ -57,8 +60,8 @@ public class IsolatedJmx
     private MBeanWrapper.InstanceMBeanWrapper wrapper;
-    private RMIClientSocketFactoryImpl clientSocketFactory;
-    private CollectingRMIServerSocketFactoryImpl serverSocketFactory;
+    private RMIClientSocketFactory clientSocketFactory;
```
@maulin-vasavada why cant we have this on `RMICloseableSocketFactory` type already? Then you do not need to cast it on line 99 and 100.

maulin-vasavada replied:
This is because the other places down the line requires `RMIClientSocketFactory` (or ServerSocketFactory) and `RMICloseableSocketFactory` type won't work there.

smiklosovic replied:
@maulin-vasavada  I am not sure how complicated it would be to rework kthat but if not possible then you should check it with `instanceof` before casting it like that. Now you assume what happens outside of this class. You assume that you will put `RMICloseableSocketFactory` into `socketFactories`.

smiklosovic commented on test/distributed/org/apache/cassandra/distributed/test/JMXSslConfigDistributedTest.java:
```diff
@@ -0,0 +1,204 @@
+/**
+ * Distributed tests for JMX SSL configuration via the system properties OR the encryption options in the cassandra.yaml.
+ */
+public class JMXSslConfigDistributedTest extends AbstractEncryptionOptionsImpl
```
It is given this test is distributed when it is in distributed package. "Distributed" is redundant.

maulin-vasavada replied:
Agree. However, it was mainly to avoid a name conflict with equivalent Unit test name `JMXSslConfigTest`.  So one of them has to differ in someway.

maulin-vasavada replied:
@smiklosovic is above reason acceptable? If so, I'll mark this comment resolved.

smiklosovic commented on test/distributed/org/apache/cassandra/distributed/test/JMXSslConfigDistributedTest.java:
```diff
@@ -0,0 +1,204 @@
+    }
+
+    @SuppressWarnings("unchecked")
+    private void configureClientSocketFactory(Map<String, Object> jmxEnv, Map<String, Object> encryptionOptionsMap) throws SSLException
```
Move this method to the end of this file. A reader just wants to see test methods first.

smiklosovic commented on test/distributed/org/apache/cassandra/distributed/test/JMXSslConfigDistributedTest.java:
```diff
@@ -0,0 +1,204 @@
+    // checkstyle: suppress below 'blockSystemPropertyUsage'
+    void setSystemTrustStore(String trustStore, String trustStorePassword)
+    {
+        System.setProperty("javax.net.ssl.trustStore", trustStore);
```
This is not a good pattern. Look for the usages of `org.apache.cassandra.distributed.shared.WithProperties` class. Ideally, they should be set in `try-catch` as that class implements `AutoCloseable` so on exit from `try-catch` these properties would be automatically reset.

maulin-vasavada replied:
I agree. I did have similar reset before, let me find out why I don't have it now.

maulin-vasavada replied:
Okay, you will see this in the next update. I didn't know about the `WithProperties` very well, now I do. Its great!

smiklosovic commented on test/distributed/org/apache/cassandra/distributed/test/JMXSslConfigDistributedTest.java:
```diff
@@ -0,0 +1,204 @@
+
+        try (Cluster cluster = builder().withNodes(1).withConfig(c -> {
+            c.with(Feature.JMX);
+            c.set("jmx_encryption_options", encryptionOptionsMap);
```
just inline options directly here, if you use WithProperties too, you'll have it pretty nice.

maulin-vasavada replied:
I didn't quite get your suggestion. Can you please suggest the change once I update the PR with other comments resolutions?

maulin-vasavada replied:
Okay, I updated the PR with addressing other comments. Now, you can make a code change suggestion.

smiklosovic commented on test/distributed/org/apache/cassandra/distributed/test/JMXSslConfigDistributedTest.java:
```diff
@@ -0,0 +1,205 @@
+    @Test
+    public void testSystemSettings() throws Throwable
+    {
+        COM_SUN_MANAGEMENT_JMXREMOTE_SSL.setBoolean(true);
```
@maulin-vasavada cant be these incorporated into WithProperties too?

maulin-vasavada replied:
Why not! Will do

maulin-vasavada replied:
I did it but only that method was using those variables. I still had to keep the `@After` cleanup for those variables because they are set as side-effect of the other tests in the same class.

smiklosovic commented on test/distributed/org/apache/cassandra/distributed/test/JMXSslConfigDistributedTest.java:
```diff
@@ -0,0 +1,205 @@
+    @Test
+    public void testInvalidKeystorePath() throws Throwable
+    {
+        ImmutableMap<String, Object> encryptionOptionsMap = ImmutableMap.<String, Object>builder()
```
just put `encryptionOptionsMap` into c.set(...), just inline it all and everywhere

maulin-vasavada replied:
oh I see. Initially I had some trouble setting this so this was left unattended. Now its fixed.

smiklosovic replied:
@maulin-vasavada you haven't inlined this, `encryptionOptionsMap` are still outside of try catch. Just make it same as you did in `testDisabledEncryptionOptions`. `testClientAuth` and `testDefaultEncryptionOptions` are not reworked either.

maulin-vasavada replied:
Fixed it.

smiklosovic commented on test/unit/org/apache/cassandra/utils/JMXSslConfigTest.java:
```diff
@@ -0,0 +1,115 @@
+
+    @After
+    public void resetJmxSslSystemProperties()
+    {
```
this whole stuff can be done with WithProperties too, no?

smiklosovic commented on src/java/org/apache/cassandra/config/EncryptionOptions.java:
```diff
@@ -182,8 +182,9 @@ static Set<String> asSet()
-                valueSet.add(key.getKeyName().toLowerCase());
+            for (ConfigKey key : values)
+            {
+                valueSet.add(key.toString().toLowerCase());
```
@maulin-vasavada use `org.apache.cassandra.utils.LocalizeString` for `toLowerCase`. This will not compile on `trunk` anymore.

maulin-vasavada replied:
For this I've to rebase it seems. I don't see that in my local.

maulin-vasavada replied:
Can we leave this until all other comments are resolved from you and other reviewers? Based on my past experience, I would take the latest trunk and reapply all the changes vs rebasing to keep commit history clear on this PR. I can pick this up as the last step after rebasing. Will leave this comment unresolved until then.

smiklosovic replied:
@maulin-vasavada could you please resolve this? Then I will build this branch in CircleCI to see where we are with it as a whole.

maulin-vasavada replied:
Yes, will get to this soon.

maulin-vasavada replied:
@smiklosovic How do you suggest I do this? I was thinking of doing following,

1. get the latest trunk to my local (already done)
2. create another branch - let us say - `CASSANDRA-18508-sensitive-jmx-ssl-config-2` from the latest trunk
3. cherry-pick the squashed commit from the branch `CASSANDRA-18508-sensitive-jmx-ssl-config` to the above branch and change this PR to be from the `CASSANDRA-18508-sensitive-jmx-ssl-config-2`

Let me know if there are better suggestions.

maulin-vasavada replied:
I managed to find a way out-

1. I got the latest trunk locally
2. I created a temp branch out of it
3. cherry-picked commit(s) from the `CASSANDRA-18508-sensitive-jmx-ssl-config` branch to the temp branch
4. resolved any conflicts in the `EncryptionOptions.java`
5. force renamed the temp branch `git branch -f -m  CASSANDRA-18508-sensitive-jmx-ssl-config`
6. fetched from myfork/`CASSANDRA-18508-sensitive-jmx-ssl-config` branch and resolved the same conflict
7. pushed the local branch to myfork to update this PR

This seem to have worked. Please review and we can move to Doug for more reviews.

smiklosovic replied:
git fetch upstream
git checkout trunk
git merge upstream/trunk
git checkout _this_branch_
git rebase trunk
((( resolve conflicts)))
git push origin _this_branch_ -f

smiklosovic replied:
@maulin-vasavada 

Your branch contains merge commits etc, that will not be deliverable on the merge. It has to be just one commit. I put all the changes into this branch.

https://github.com/instaclustr/cassandra/commit/1d172ce58f2587ab521696e6037ae8ab6bc16c94

you can hard reset all your changes in your branch and cherry-pick that one commit and force-push it to  your remote to be on par with it.

Please create one commit for this branch so everybody has easier life. 

The way I did that is that I did "git rebase trunk" while on this branch you just pushed. There was one conflict I resolved. After that I just force-pushed and got that commit above. You can do the same.

smiklosovic replied:
@maulin-vasavada 

https://app.circleci.com/pipelines/github/instaclustr/cassandra/4958/workflows/134abde9-5c0d-47f6-82d5-d02f9be48078


Fails on 

````
com.datastax.driver.core.exceptions.ServerError: An unexpected error occurred server side on localhost/127.0.0.1:46255: org.yaml.snakeyaml.error.YAMLException: Failed calling getter for property 'optional' on class org.apache.cassandra.config.EncryptionOptions

	Suppressed: java.lang.RuntimeException: Error calling get() on jmx_encryption_options.optional of class java.lang.Boolean
		at org.apache.cassandra.config.Properties$AndThen.get(Properties.java:191)
		... 28 common frames omitted
Caused by: java.lang.IllegalStateException: EncryptionOptions.applyConfig must be called first
	at org.apache.cassandra.config.EncryptionOptions.ensureConfigApplied(EncryptionOptions.java:368)
	at org.apache.cassandra.config.EncryptionOptions.getOptional(EncryptionOptions.java:413)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.apache.cassandra.config.DefaultLoader$MethodPropertyPlus.get(DefaultLoader.java:107)
	... 29 common frames omitted
````


```
  ✕ j17_unit_tests                                  15m 17s
      org.apache.cassandra.db.virtual.SettingsTableTest testTransparentEncryptionOptionsOverride
      org.apache.cassandra.db.virtual.SettingsTableTest testArray
      org.apache.cassandra.db.virtual.SettingsTableTest testSelectAll
      org.apache.cassandra.db.virtual.SettingsTableTest testSelectEmpty
      org.apache.cassandra.db.virtual.SettingsTableTest testAuditOverride
      org.apache.cassandra.db.virtual.SettingsTableTest testSelectOverride
      org.apache.cassandra.db.virtual.SettingsTableTest testSelectPartition
      org.apache.cassandra.db.virtual.SettingsTableTest testEncryptionOverride
      org.apache.cassandra.config.PropertiesTest backAndForth
```

maulin-vasavada replied:
@smiklosovic I fixed that test.

maulin-vasavada replied:
Once everything looks I'll squash commits.

smiklosovic commented on src/java/org/apache/cassandra/utils/jmx/AbstractJmxSocketFactory.java:
```diff
@@ -0,0 +1,161 @@
+     * @param jmxEncryptionOptions for the SSL communication
+     * @throws SSLException if fails to configure the SSL based server socket factory
+     */
+    abstract public void configureSslServerSocketFactoryBasedOnEncryptionOptions(Map<String, Object> env, InetAddress serverAddress,
```
@maulin-vasavada I am sorry I am so picky here but ... The way these methods should be called is that I will know _why_ I am calling it. It is cool that I know that I am calling a method which configures ssl server factory based on encryption options, but _why_ do I do that? What is the outcome of me doing this? In the implementation I see that `JmxSslRMIServerSocketFactory` is going to be created. That tells that I will configure connection factory backed by JMX. So why  not to call this method `configureJMXBackedSSLServerFactory`?

Do you see where I am going with this? 

I think that all factories will be jmx enabled after all so maybe my commentary is not completely accurate neverthless that method should tell what is the intent of me call it. Here I see that I am indeed passing EncryptionOptions into that so of course it will be "based on EncryptionOptions", I do not need to have this information duplicated. I just need to know _why_ I should care calling it.

smiklosovic replied:
Also, I am not completely sure why we have just one "configure" method in IJmxSocketFactory but here we are adding more of them to be implemented which are called in the method of the interface we implement in this abstract class.

Can not we have just `AbstractJmxSocketFactory` where basic "configure" method is implemented and all other methods are abstract? Then you would just extend this abstract class twice for your needs.

In other words, I just dont see what is IJMXSocketFactory good for. Just implement it in abstract class and leave other methods abstract.

The way how it is now is that implementators of IJMXSocketFactory can create a class which implements configure method but they do not need to call any abstract method here. If we removed interface and we had one abstract class with configure method implemented in such a way that it would contain other methods which are abstract here then whoever extends this abstract class _has to_ implement them.

maulin-vasavada replied:
@smiklosovic So doesn't current class hierarchy do that exactly? `IJmxSocketFactory <- AbstractJmxSocketFactory <- DefaultJmxSocketFactory / IsolatedJmxSocketFactory`. Abstract class does have the basic configure method implemented and other are abstract which are implemented by the Default factory AND IsolatedJmx factory.

![image](https://github.com/user-attachments/assets/30629157-7678-456c-9d4d-98991c7a1d11)

smiklosovic replied:
@maulin-vasavada  What do we need `IJmxSocketFactory` for? Why can you not have just `AbstractJmxSocketFactory` with implemented `public Map<String, Object> configure(InetAddress serverAddress, boolean localOnly, EncryptionOptions jmxEncryptionOptions)`  as now and keep other methods abstract.

maulin-vasavada replied:
Interesting question. My mind is tuned to always have an interface and have abstract classes as in-between code optimizations. Let me think a little bit and get back. From technical perspective we may do without it.

maulin-vasavada replied:
Okay @smiklosovic I've removed the `IJmxSocketFactory` interface. My thinking is - since we do not have a plan for the pluggable implementation for the Jmx SocketFactory (as current usage is directly creating the `Default` or `Isolated` Jmx socket factories), we really don't have a strong reason to have that interface. In future it could change but for now the scope is clearer with just the direct implementations for the abstract class. Again the abstract class is created mainly to consolidate prior duplicated logic (as discussed in the jira ticket with Doug).

I also renamed the `abstract configure` methods which I ended up suffixing with `BasedOnSystemConfig` etc. Please review that again.

smiklosovic replied:
@maulin-vasavada I see the interface removed but I do not see the method renamed as you said. Have you pushed?

maulin-vasavada replied:
Yes I did push. I was referring to these methods in the abstract class  -
![image](https://github.com/user-attachments/assets/4ef53b67-f0dd-4ec3-87e2-af4c4150ed77)

smiklosovic replied:
okay

smiklosovic commented on test/distributed/org/apache/cassandra/distributed/impl/RMISslClientSocketFactoryImpl.java:
```diff
@@ -0,0 +1,146 @@
+        return socket;
+    }
+
+    private Socket createSslSocket(int port) throws IOException {
```
@maulin-vasavada please put braces on new lines everywhere. Please be consistent with it.

maulin-vasavada replied:
sorry about it. I forgot to reformat this.

smiklosovic commented on test/distributed/org/apache/cassandra/distributed/impl/JmxTestClientSslSocketFactory.java:
```diff
@@ -0,0 +1,72 @@
+    @Override
+    public Socket createSocket(String host, int port) throws IOException
+    {
+        // Retrieve the SSLSocketFactory
```
the comments in this method are unnecessary, the code is pretty much self evident.

smiklosovic commented on src/java/org/apache/cassandra/utils/RMICloseableSocketFactory.java:
```diff
@@ -0,0 +1,33 @@
+/**
+ * This represents closeable RMI Socket factory. It could be implemented for client and/or server socket factories.
+ */
+public interface RMICloseableSocketFactory
```
@maulin-vasavada you can rework this to abstract class extending `RMISocketFactory` and then all current implementations extending `RMISocketFactory` would extend `RMICloseableSocketFactory` instead. That way you can set clientFactory below to be RMICloseableSocketFactory and you do not need to cast anything.

maulin-vasavada replied:
Will look into it.

maulin-vasavada replied:
There is no common ground between `RMIClientSocketFactory` and `RMIServerSocketFactory` from `java.rmi.server`. So to follow your suggestion I'd have to create `RMICloseableClientSocketFactory` and `RMICloesableServerSocketFactory` interfaces and then ultimately avoid the casting you are pointing to.

smiklosovic commented on test/distributed/org/apache/cassandra/distributed/test/JMXSslConfigDistributedTest.java:
```diff
@@ -0,0 +1,202 @@
+                                                                "javax.net.ssl.keyStorePassword", (String) validKeystore.get("keystore_password"))
+        )
+        {
+            ImmutableMap<String, Object> encryptionOptionsMap = ImmutableMap.<String, Object>builder().putAll(validKeystore)
```
@maulin-vasavada inline please into try-catch

maulin-vasavada replied:
@smiklosovic The challenge is that we have to use this map later in the block within try catch. So I need to have a variable to hold the values unlike the inline setting in the other method. Please check the `line # 90` and you will see what I mean.

smiklosovic commented on test/distributed/org/apache/cassandra/distributed/test/JMXSslConfigDistributedTest.java:
```diff
@@ -0,0 +1,202 @@
+                                                                "javax.net.ssl.keyStorePassword", (String) validKeystore.get("keystore_password"))
+        )
+        {
+            ImmutableMap<String, Object> encryptionOptionsMap = ImmutableMap.<String, Object>builder().putAll(validKeystore)
```
same

maulin-vasavada replied:
same challenge here as mentioned before. I need to use the map variable later.

smiklosovic commented on test/distributed/org/apache/cassandra/distributed/impl/IsolatedJmx.java:
```diff
@@ -84,14 +87,17 @@ public void startJmx() {
+            // However if the `jmxEncryptionOptions` are provided or JMX SSL configuration is set it will configure
+            // the socket factories appropriately.
+            Map<String, Object> socketFactories = new IsolatedJmxSocketFactory().configure(addr, true, jmxEncryptionOptions);
+            serverSocketFactory = (RMICloseableServerSocketFactory) socketFactories.get(RMIConnectorServer.RMI_SERVER_SOCKET_FACTORY_ATTRIBUTE);
```
@maulin-vasavada do we still need to cast?

maulin-vasavada replied:
yes because the return map has `Object` type so we have to cast.

maulin-vasavada committed:
Fixing failing unit test

maulin-vasavada commented:
@smiklosovic just a clarification that [27cd310](https://github.com/apache/cassandra/pull/3638/commits/27cd310e5e012bf96e7c6e26b0b3993c9214f9e6) in my commit history is not trunk merge or anything. None-the-less, I'll follow your suggested steps once one more reviewer comments are addressed.

maulin-vasavada committed:
Refactoring as per older PR comments - removing JmxSslRMIServerSocketFactory and moving jmx unit tests in separate package

maulin-vasavada committed:
adding back the accidentally removed comments

maulin-vasavada committed:
Fixing a test failure

maulin-vasavada commented:
@smiklosovic I've fixed the latest failing test and pushed the commit. However, now I am running ALL unit tests locally to validate if anything fails due to the jmx encryption options. It would take some time to finish it though. Will post what I find.

maulin-vasavada committed:
Logging for Assertion error better in the JMX tests

maulin-vasavada committed:
Changing the log to be of ERROR level

maulin-vasavada commented:
@smiklosovic I moved the JMX test method to a utility class and added `ERROR` level log for each error before adding it to the suppressed list. Let us see if this helps us identify flakiness in the test.

maulin-vasavada committed:
Reverting accidental change

maulin-vasavada committed:
Using JMXTestsUtil in other JMX tests

JeetKunDoug commented on src/java/org/apache/cassandra/config/DatabaseDescriptor.java:
```diff
@@ -1306,7 +1309,14 @@ public static void applySslContext()
+            SSLFactory.validateSslContext("JMX transport", conf.jmx_encryption_options, conf.jmx_encryption_options.getClientAuth(), true);
             SSLFactory.initHotReloading(conf.server_encryption_options, conf.client_encryption_options, false);
+            /*
+            For JMX SSL, the hot reloading of the SSLContext is out of scope for CASSANDRA-18508.
```
Did we create a JIRA to track this, as the lack of hot-reloading of SSL certificates is going to be a real issue for folks who use short-lived SSL certs.

maulin-vasavada replied:
Yes, I did - [CASSANDRA-20067](https://issues.apache.org/jira/browse/CASSANDRA-20067)

JeetKunDoug commented on src/java/org/apache/cassandra/utils/RMICloseableServerSocketFactory.java:
```diff
@@ -0,0 +1,34 @@
+/**
+ * This represents closeable RMI Server Socket factory.
+ */
+public interface RMICloseableServerSocketFactory extends RMIServerSocketFactory
```
This is essentially:
```suggestion
public interface RMICloseableServerSocketFactory extends RMIServerSocketFactory, AutoCloseable
```
I am trying to figure out if there's a reason we _wouldn't_ want to also extend `AutoCloseable` rather than adding a custom `close()` method... while I can't think of a reason why you'd want to use one in a "try-with-resources" block at this time, it also doesn't hurt anything to _allow it_ in the future, and by not doing it now we're artificially limiting the ability to do so. Also, there are some frameworks that can check if something is AutoCloseable and close it on shutdown/cleanup (again, nothing today seems to do this but it's possible we may want to some day).

JeetKunDoug replied:
FYI - this would go for the Client socket factory as well.

maulin-vasavada replied:
Makes sense. Will do the changes. Since I've to remove the `close()` method in this case, will add to my changes instead of picking up the commit suggestion.

JeetKunDoug commented on test/distributed/org/apache/cassandra/distributed/impl/CollectingSslRMIServerSocketFactoryImpl.java:
```diff
@@ -0,0 +1,157 @@
+    private final String[] enabledCipherSuites;
+    private final String[] enabledProtocols;
+    private final boolean needClientAuth;
+    private SSLContext sslContext;
```
If we change this field to be:
```suggestion
    private final SSLSocketFactory sslSocketFactory;
```
We can then initialize it using the `sslContext` passed in to the constructor that uses it, _or_ use `getDefaultSSLSocketFactory` in the other, which gets rid of the ternary operator in `createSslServerSocket`. Do you envision a need to hold on to the `sslContext` itself, or could we do this in stead?

maulin-vasavada replied:
Ack. Will pick it up.

JeetKunDoug commented on test/distributed/org/apache/cassandra/distributed/impl/IsolatedJmx.java:
```diff
@@ -137,6 +143,50 @@ public void startJmx() {
+        } else
+        {
+            return null;
+        }
```
I tend to try to invert these kinds of things, which generally allows a quick return on `null` and then you don't actually need the extra indentation/else statement:
```suggestion
    private EncryptionOptions getJmxEncryptionOptions(Map<String,Object> encryptionOptionsMap)
    {
        if (encryptionOptionsMap == null)
        {
            return null;
        }
        EncryptionOptions jmxEncryptionOptions = new EncryptionOptions();
        String[] cipherSuitesArray = (String[])encryptionOptionsMap.get(EncryptionOptions.ConfigKey.CIPHER_SUITES.toString());
        if (cipherSuitesArray != null )
        {
            jmxEncryptionOptions = jmxEncryptionOptions.withCipherSuites(cipherSuitesArray);
        }
        List<String> acceptedProtocols = (List<String>)encryptionOptionsMap.get(EncryptionOptions.ConfigKey.ACCEPTED_PROTOCOLS.toString());
        if (acceptedProtocols != null )
        {
            jmxEncryptionOptions = jmxEncryptionOptions.withAcceptedProtocols(acceptedProtocols);
        }

        Boolean requireClientAuthValue = (Boolean) encryptionOptionsMap.get(EncryptionOptions.ConfigKey.REQUIRE_CLIENT_AUTH.toString());
        EncryptionOptions.ClientAuth requireClientAuth = requireClientAuthValue == null ?
                                                         EncryptionOptions.ClientAuth.NOT_REQUIRED :
                                                         EncryptionOptions.ClientAuth.from(String.valueOf(requireClientAuthValue));
        Object enabledOption = encryptionOptionsMap.get(EncryptionOptions.ConfigKey.ENABLED.toString());
        boolean enabled = enabledOption != null ? (Boolean)encryptionOptionsMap.get(EncryptionOptions.ConfigKey.ENABLED.toString()) : false;

        //CASSANDRA-18508 NOTE- We do not populate sslContextFactory configuration here for tests, it could be enhanced
        jmxEncryptionOptions = jmxEncryptionOptions
               .withKeyStore((String)encryptionOptionsMap.get(EncryptionOptions.ConfigKey.KEYSTORE.toString()))
               .withKeyStorePassword((String)encryptionOptionsMap.get(EncryptionOptions.ConfigKey.KEYSTORE_PASSWORD.toString()))
               .withTrustStore((String)encryptionOptionsMap.get(EncryptionOptions.ConfigKey.TRUSTSTORE.toString()))
               .withTrustStorePassword((String)encryptionOptionsMap.get(EncryptionOptions.ConfigKey.TRUSTSTORE_PASSWORD.toString()))
               .withRequireClientAuth(requireClientAuth)
               .withEnabled(enabled);
        return jmxEncryptionOptions;
    }
```

JeetKunDoug reviewed (commented):
Thanks again for this important contribution. About half way through the review, and I have a few comments/suggestions but things look great.

I'm submitting my comments to date to unblock you on replying/making any changes that make sense.

maulin-vasavada committed:
Update test/distributed/org/apache/cassandra/distributed/impl/IsolatedJmx.java

Co-authored-by: Doug Rohrer <682821+JeetKunDoug@users.noreply.github.com>

maulin-vasavada committed:
Addressing latest PR comments

maulin-vasavada committed:
Merge branch 'CASSANDRA-18508-sensitive-jmx-ssl-config' of https://github.com/maulin-vasavada/cassandra into CASSANDRA-18508-sensitive-jmx-ssl-config

maulin-vasavada committed:
Update IsolatedJmx.java

JeetKunDoug commented on test/distributed/org/apache/cassandra/distributed/impl/IsolatedJmxSocketFactory.java:
```diff
@@ -0,0 +1,96 @@
+    private static final Logger logger = LoggerFactory.getLogger(IsolatedJmxSocketFactory.class);
+
+    @Override
+    public void configureLocalSocketFactory(Map<String, Object> env, InetAddress serverAddress)
```
NIT: This should be named `configureSslServerSocketFactories` since it configured both.

JeetKunDoug commented on test/distributed/org/apache/cassandra/distributed/impl/JmxTestClientSslContextFactory.java:
```diff
@@ -0,0 +1,145 @@
+import org.apache.cassandra.io.util.File;
+
+/**
+ * Simplied and independent version of {@link org.apache.cassandra.security.FileBasedSslContextFactory} for
```
NIT: Spelling - assuming this was meant to be `Simplified`?
```suggestion
 * Simplified and independent version of {@link org.apache.cassandra.security.FileBasedSslContextFactory} for
```

JeetKunDoug commented on test/distributed/org/apache/cassandra/distributed/impl/JmxTestClientSslContextFactory.java:
```diff
@@ -0,0 +1,145 @@
+    // keystore is not needed when the JMX server does not require client-auth
+    protected String keystore;
+    // keystore could be null in case JMX server does not require client-auth
+    protected String keystore_password;
```
NIT: both of these can be final as well, and all of the member variables can be private.

maulin-vasavada committed:
moving JMX ssl distributed test to jmx package

JeetKunDoug commented on test/distributed/org/apache/cassandra/distributed/impl/JmxTestClientSslContextFactory.java:
```diff
@@ -0,0 +1,145 @@
+    protected Boolean getBoolean(String key)
+    {
+        return (Boolean) this.parameters.get(key);
+    }
```
both `getBoolean` methods are unused.

JeetKunDoug commented on test/distributed/org/apache/cassandra/distributed/impl/RMISslClientSocketFactoryImpl.java:
```diff
@@ -0,0 +1,148 @@
+
+    private static synchronized SocketFactory getDefaultClientSocketFactory()
+    {
+        if (defaultSocketFactory == null)
```
Is there a reason we do this lazy initialization here? it's always initialized to the same value, it isn't changed externally with a setter or anything like that. Further, `SSLSocketFactory.getDefault()` is _already synchronized and lazy_ so we could just call it directly in the one place this method is used.

JeetKunDoug commented on test/distributed/org/apache/cassandra/distributed/impl/RMISslClientSocketFactoryImpl.java:
```diff
@@ -0,0 +1,148 @@
+            }
+        }
+        return sslSocket;
+    }
```
A few suggestions for simplifying this:

1) Extract the string splitting into a method, and maybe just use a regex:

```
    private static final Pattern COMMA_SPLITTER = Pattern.compile(",");
    private String[] splitCommaSeparatedString(String stringToSplit)
    {
        if (stringToSplit == null)
            return null;
        return COMMA_SPLITTER.split(stringToSplit);
    }
```

2) Change the two string members (`enabledCipherSuites` and `enabledProtocols`) to `String[]`s rather than just strings.

3) Split them in the constructor rather than every time through this method
4) Then you can just check for null and, if they are null, call the approbate `set` method directly.

JeetKunDoug commented on test/distributed/org/apache/cassandra/distributed/test/AbstractEncryptionOptionsImpl.java:
```diff
@@ -70,11 +70,16 @@ public class AbstractEncryptionOptionsImpl extends TestBaseImpl
                                                                     "truststore_password", validTrustStorePassword);
 
+    // Configuration for a valid PEM based keystore
+    final static Map<String, Object> validPEMKeystore = ImmutableMap.of("ssl_context_factory", ImmutableMap.of("class_name", "org.apache.cassandra.security.PEMBasedSslContextFactory"),
```
This appears to be unused - we should either add a test that uses it, or remove it. I would guess we want to add a test somewhere?

maulin-vasavada replied:
Will remove it since for the testing it relies on javax properties for the key/trust store which we can't use for PEM formatted keystores.

JeetKunDoug commented on test/distributed/org/apache/cassandra/distributed/test/jmx/JMXSslConfigDistributedTest.java:
```diff
@@ -0,0 +1,202 @@
+        COM_SUN_MANAGEMENT_JMXREMOTE_SSL_ENABLED_PROTOCOLS.reset();
+        COM_SUN_MANAGEMENT_JMXREMOTE_SSL_ENABLED_CIPHER_SUITES.reset();
+        JAVAX_RMI_SSL_CLIENT_ENABLED_PROTOCOLS.reset();
+        JAVAX_RMI_SSL_CLIENT_ENABLED_CIPHER_SUITES.reset();
```
This code is repeated several times in different tests. Please refactor it to JmxTestsUtil.java and call it from the different tests that use it.

JeetKunDoug commented on test/distributed/org/apache/cassandra/distributed/test/jmx/JMXSslConfigDistributedTest.java:
```diff
@@ -0,0 +1,202 @@
+            {
+                Map<String, Object> jmxEnv = new HashMap<>();
+                configureClientSocketFactory(jmxEnv, encryptionOptionsMap);
+                // Invoke the same code vs duplicating any code from the JMXGetterCheckTest
```
NIT: I think you can remove these comments now that this code is refactored into the test utility. If not, at least please update the class name at the end of the comments to reference the place where the code lives.

JeetKunDoug commented on test/distributed/org/apache/cassandra/distributed/test/AbstractEncryptionOptionsImpl.java:
```diff
@@ -70,11 +70,16 @@ public class AbstractEncryptionOptionsImpl extends TestBaseImpl
 
     // Base configuration map for a valid keystore that can be opened
-    final static Map<String,Object> validKeystore = ImmutableMap.of("keystore", validKeyStorePath,
+    protected final static Map<String,Object> validKeystore = ImmutableMap.of("keystore", validKeyStorePath,
```
NIT: This is a bit more than a valid _keystore_ - maybe call it `validFileBasedStores`? That way if we keep `validPEMKeystore` we can rename it `validPEMBasedStores`?

maulin-vasavada replied:
will be in the next commit

JeetKunDoug commented on test/unit/org/apache/cassandra/utils/jmx/JMXSslConfigTest.java:
```diff
@@ -0,0 +1,120 @@
+        // The below properties are set as side effect of other properties when tests run. Hence, these are reset here
+        // vs using try-with-resouces block
+        JAVAX_RMI_SSL_CLIENT_ENABLED_PROTOCOLS.reset();
+        JAVAX_RMI_SSL_CLIENT_ENABLED_CIPHER_SUITES.reset();
```
The fact that some of this is done here, and some with the `WithProperties` below let me to think there's a missing feature in `WithProperties` - if you add the following method there:

```java
    public WithProperties preserve(CassandraRelevantProperties prop)
    {
        String previous = prop.getString(); // because all properties are strings
        rollback.add(previous == null ? prop::clearValue : () -> prop.setString(previous));
        return this;
    }
```

You can specify all of the properties you need to roll back in the single `WithProperties` block (and maybe could use that in other places where we're rolling back other properties?)

I can take a look at this separately if you're not comfortable making this change, but it would be a nice addition to `WithProperties`

JeetKunDoug commented on test/unit/org/apache/cassandra/utils/jmx/JMXSslDefaultEncryptionOptionsTest.java:
```diff
@@ -0,0 +1,113 @@
+        COM_SUN_MANAGEMENT_JMXREMOTE_SSL_ENABLED_PROTOCOLS.reset();
+        COM_SUN_MANAGEMENT_JMXREMOTE_SSL_ENABLED_CIPHER_SUITES.reset();
+        JAVAX_RMI_SSL_CLIENT_ENABLED_PROTOCOLS.reset();
+        JAVAX_RMI_SSL_CLIENT_ENABLED_CIPHER_SUITES.reset();
```
Same comment about pulling this into the JMXTestsUtil class.

maulin-vasavada replied:
This comment I missed to understand. I made the `WithProperties` changes. Do you suggest to move any of those to the JMXTestsUtil OR were you suggesting that if we keep this `@After` tear down methods we should move it to JMXTestsUtil?

JeetKunDoug commented on test/unit/org/apache/cassandra/utils/jmx/JMXSslDefaultEncryptionOptionsTest.java:
```diff
@@ -0,0 +1,113 @@
+ * The default configurtion means keystore/truststore configured with file paths and using {@code DefaultSslContextFactory}
+ * to initialize the SSLContext.
+ */
+public class JMXSslDefaultEncryptionOptionsTest
```
The `Default` name here really confused me for a while... can we call this test something like `JMXSslConfiguredWithYamlFileOptionsTest` or something more descriptive, as the _default_ is really no JMX encryption. It seems like `JMXSslConfigTest#testLocalJmxServer` represents the "default" configuration right?

maulin-vasavada replied:
This confusion is because of the perception I had when I named it. I intended to say - this test validates default `jmx_encryption_options` when it is not configured in the yaml. I tried to explain that in the comments for the class but it still managed to confuse you. However, when I read your comment, I can feel the confusion from a different point-of-view and if I rename the class/method with your suggestion, it may avoid any confusion. Let me rename it.

JeetKunDoug commented on test/unit/org/apache/cassandra/utils/jmx/JMXSslDisabledEncryptionOptionsTest.java:
```diff
@@ -0,0 +1,110 @@
+    @After
+    public void resetJmxSslSystemProperties()
+    {
+        COM_SUN_MANAGEMENT_JMXREMOTE_SSL.reset();
```
Same comment about moving these into the utility class.

JeetKunDoug commented on test/unit/org/apache/cassandra/utils/jmx/JMXSslDisabledEncryptionOptionsTest.java:
```diff
@@ -0,0 +1,110 @@
+     * 3. jmx_encryption_options in the cassandra.yaml
+     */
+    @Test
+    public void testAbsenceOfAllJmxSslConfigs() throws SSLException
```
The comments and the setup don't really match in this case - the _absense_ of all JMX SSL configuration should be no JMX config in cassandra.yaml, right? But this class loads a config that has those settings, but disabled.

maulin-vasavada replied:
The challenge is because `jmx_encryption_options` get defaults when we don't specify it in cassandra.yaml. So I meant to say that - this test validates use-case when none of the below are configured-

1. JMX is turned off with `com.sun.management.jmxremote.ssl` system property
2. system configuration for the JMX
3. `jmx_encryption_options` in cassandra.yaml

I have tried to rename the method and update the documentation. Please review if the next change makes more sense.

JeetKunDoug reviewed (changes requested):
Some more comments/suggestions/quesitons - nothing major, but some test-related stuff especially is a bit confusing for me, so I'd like to get that cleaned up/clarified (naming is hard :P )

maulin-vasavada committed:
Update test/distributed/org/apache/cassandra/distributed/impl/JmxTestClientSslContextFactory.java

Co-authored-by: Doug Rohrer <682821+JeetKunDoug@users.noreply.github.com>

maulin-vasavada committed:
addressing more PR review comments

maulin-vasavada committed:
Moving common properties settings in tests to util class

maulin-vasavada committed:
addressing latest PR comments

maulin-vasavada committed:
removing unnecessary code comment

maulin-vasavada committed:
Adding a test for WithProperties

maulin-vasavada committed:
Replacing StringTokenizer with Pattern for CSV separation

maulin-vasavada committed:
documentation changes

maulin-vasavada committed:
Update security.adoc

maulin-vasavada committed:
Update security.adoc
